language/move-vm/runtime/src/session.rs

```diff
@@ -93,38 +117,68 @@ impl<'r, 'l, S: MoveResolver> Session<'r, 'l, S> {
     /// used to manufacture `signer`'s or `Coin`'s from raw bytes. It is the respmsibility of the
     /// caller (e.g. adapter) to ensure that this power is useed responsibility/securely for its use-case.
     pub fn execute_function_for_effects<V>(
-        mut self,
+        self,
         module: &ModuleId,
         function_name: &IdentStr,
         ty_args: Vec<TypeTag>,
         args: Vec<V>,
         gas_status: &mut GasStatus,
     ) -> ExecutionResult
+    where
+        V: Borrow<[u8]>,
+    {
+        let no_ext = NativeContextExtensions::default();
+        self.execute_function_for_effects_with_extensions(
+            module,
+            function_name,
+            ty_args,
+            args,
+            gas_status,
+            no_ext,
+        )
+        .1
+    }
+
+    /// Same as `execute_function_for_effects`, but with native context extensions.
+    pub fn execute_function_for_effects_with_extensions<V>(
+        mut self,
+        module: &ModuleId,
+        function_name: &IdentStr,
+        ty_args: Vec<TypeTag>,
+        args: Vec<V>,
+        gas_status: &mut GasStatus,
+        mut extensions: NativeContextExtensions,
+    ) -> (NativeContextExtensions, ExecutionResult)
     where
         V: Borrow<[u8]>,
     {
         let gas_budget = gas_status.remaining_gas().get();
+
         let execution_res = self.runtime.execute_function_for_effects(
             module,
             function_name,
             ty_args,
             args,
             &mut self.data_cache,
             gas_status,
+            &mut extensions,
         );
         let gas_used = gas_budget - gas_status.remaining_gas().get();
         match execution_res {
             Ok((return_values, mutable_ref_values)) => match self.finish() {
-                Ok((change_set, events)) => ExecutionResult::Success {
-                    change_set,
-                    events,
-                    return_values,
-                    mutable_ref_values,
-                    gas_used,
-                },
-                Err(error) => ExecutionResult::Fail { error, gas_used },
+                Ok((change_set, events)) => (
+                    extensions,
+                    ExecutionResult::Success {
+                        change_set,
+                        events,
+                        return_values,
+                        mutable_ref_values,
+                        gas_used,
+                    },
+                ),
+                Err(error) => (extensions, ExecutionResult::Fail { error, gas_used }),
             },
-            Err(error) => ExecutionResult::Fail { error, gas_used },
+            Err(error) => (extensions, ExecutionResult::Fail { error, gas_used }),
```

Comment:
I disagree. Extensions can record information about failure, and therefore should be passed back.

Reply:
What does that mean? Like no other part of Move has the ability to do that (outside of maybe gas? if you squint a bit), you either get an abort code, or produce a changeset

Reply:
You gave the answer in your reply already: gas is an example. An extension may record info similar as GasStatus. It is a generic concept.

Reply:
Not 100% sure if it applies here, but I think here's another potential application to consider: events.

Different platforms may have different semantics for events. For example, EVM drops all events if the transaction gets reverted, while Diem/Move still emits them in case of failure. As Move evolve into a language targeting multiple platforms, I wonder if we would benefit from some of the extra flexibilities here.

Reply:
Backing up a bit here, its not about &mut or not. It's about what effects are easily observable in the case that the Move program aborts. Currently no changes to the outer state are observable because all changes are local, because the session has a local view of changes to storage, and reveals them only through finish(). It has no ability to write to storage because the DataStore API doesn't expose it.

My worry is that the way you have setup the extensions would be like if the session and the interpreter wrote to storage directly, and then it was up to the adapter to back to those changes out in the case of failure. Instead of what we do now, which is the VM never touches your data directly, and instead produces a set of changes locally and returns them with the changeset

This is exactly what @vgao1996 was referring to:

> IIRC, @tnowacki and I had a related discussion when we were working on extensible native functions, and one concern that was brought up was that the mutation of such external environments breaks the transactional nature of the Move VM. With that being said I think things are different now with more discussions around giving more control to the adapter layer, and that "being transactional" is a property the adapter (not necessarily the Move VM) needs to hold.

I agree with @vgao1996  that we are more in the world now where the adapter has some more responsibilities. But I still want to make sure that doing the wrong thing is impossible/difficult. And doing the right thing is obvious/easy. 

I worry with the current PR here, it would be very difficult for an adapter to keep a transactional view of any data passed around these extensions

Reply:
Extensions are -- as the name says -- extensions to the current session execution model. I would expect that a well designed extension would follow exactly the same model as the session: it can be created, then a sequence of mutations happens (via native functions), and then there is a finalize call which consumes the extension and computes the extension changeset. There can be, however, also finalize_on_failure, and that is what this API tries supports.

I'm not in favor of restricting APIs based on speculation about the semantics of future extensions. There maybe many reasons why want to have finalize_on_failure for extensions. Beside some kind of custom gas, there can be some debugging info which one wants to pick from an extension. Why should we restrict this? I did not hear any convincing arguments. And I don't really understand why we spent that much time on this little technical detail. As said, we can always come back and change if we feel it's important enough. Let's build the first actual extensions and possibly iterate, instead of being bogged down on this small technical detail.

Reply:
Missed @vgao1996's comment here

> while Diem/Move still emits them in case of failure.

That seems... odd. I did not know it did that, lol. Feels weird to have an event like "PaymentSent" or something just to load it up and see no such transaction. (we can take the discussion offline as it is not pertinent to this PR, but do you recall why that behavior exists, @vgao1996?)

> instead of being bogged down on this small technical detail.

I'm not trying to do that @wrwg, and I apologize if it is coming across that way. To summarize, the concern I have is that the modifications to this native function storage will get **accidentally** misused by an adapter. I want to make sure that any non-transactional behavior, such as Diem events (apparently), is done so intentionally and purposefully. 

I am trying to make sure the design fits into that line of thinking. My hope is that by passing in an owned copy of the extension to all functions, and only getting it back through endpoints that are designed to be run for side-effects. That there would be some more apparently intentionality in the design. Of course you can always sidestep this by wrapping data in `RC`s, or something similar, inside of the native extension, but you have to at least do some work to do that, thus making it an intentional choice. 
And obviously here we aren't doing much by forcing to have an owned value, and I wish there was a better solution (i've thought of doing things like requiring `Send` or `Sync` on the value but I don't think that helps too much really), but I haven't come up with much other than just the owned value thing.

Reply:
>  To summarize, the concern I have is that the modifications to this native function storage will get accidentally misused by an adapter. 

I think this is overdesign at the wrong place. Adapters are not everyday code and there aren't dozens of them. People which write adapters can shoot into there foot in many ways and need to know what they are doing. Restricting them unnecessarily (like you are suggesting here) is not helpful for the future of the evolution of Move. There have been multiple examples brought to your attention where having the extension also after failure is useful: the event semantics, debugging, additional gas model for natives, etc.

Besides, this is a new experimental API and we can iterate as we evolve the extension mechanism. It does not make sense to block this PR based on the arguments you brought so far,

Reply:
... for the sake of moving forward and look at more interesting things, I will adopt the PR to what you suggest, while I disagree anyway. Stay tuned for the next commit.

Reply:
The change you are asking for has been pushed now.